modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequest.java

```diff
@@ -41,23 +43,34 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Collectors;
 
 public class Netty4HttpRequest implements HttpRequest {
-    private final FullHttpRequest request;
-    private final BytesReference content;
     private final HttpHeadersMap headers;
     private final int sequence;
+    private final AtomicBoolean released;
+    private final FullHttpRequest request;
+    private final boolean pooled;
+    private BytesReference content;
 
     Netty4HttpRequest(FullHttpRequest request, int sequence) {
+        this(request, sequence, new AtomicBoolean(false));
+    }
+
+    private Netty4HttpRequest(FullHttpRequest request, int sequence, AtomicBoolean released) {
+        this(request, new HttpHeadersMap(request.headers()), sequence, released, request.content() instanceof UnpooledHeapByteBuf == false,
```

Comment:
I wonder how reliable this check is. In particular, how do we make sure this is not just a NOOP and we always return false? Can we have a test somewhere for this?

Reply:
Just tried adding a test for this and it turned out to be a little trickier than expected. We currently only expose setting the unpooled allocator via a system property so this would require some additional code changes to make it testable.

I'm fine dropping this as well though, the unpooled allocator is pretty rare and it may not be worth the complexity optimizing for it here (since we're only optimizing the small request size case anyway). I just figured, since doing this check constrained to just `UnpooledHeapByteBuf`
was essentially a freebie and it worked in manual testing I'd go for it? :)

Reply:
@tbrooks8 thoughts on how to test this and/or whether to drop this?

Could we assert e.g. if `NettyAllocator.useUnpooled()` is true, that we then have an instance of `UnpooledHeapByteBuf` here?

Reply:
> Could we assert e.g. if NettyAllocator.useUnpooled() is true, that we then have an instance of UnpooledHeapByteBuf here?

Unfortunately no, because larger messages will often show up as `CompositeByteBuf` (which we'd have to deep-inspect to make sure they don't contain any pooled buffers).  What a test for this would want to do, is to check that the unpooled allocator produces "some" instances of `UnpooledHeapByteBuf` here.
-> If this makes us uncomfortable I'd suggest just dropping it for now and revisiting optimizing the unpooled case later (if we care).

Reply:
I ditched this check in 9ab9b4a now, realizing that it may not be safe in all cases anyway because `UnpooledHeapByteBuf` isn't `final` so if a child overrides the `release` method for whatever reason we'd be in trouble potentially.

I think we can do a better job here anyway up-stream by simply making use of the knowledge that we're running the unpooled allocator. I'd optimize this in a follow-up.

Can you take another look @ywelsch ? :)